shared/naturalcrit/codeEditor/codeEditor.less

```diff
@@ -1,10 +1,25 @@
 @import (less) 'codemirror/lib/codemirror.css';
 @import (less) 'codemirror/addon/fold/foldgutter.css';
+@import (less) 'codemirror/addon/search/matchesonscrollbar.css';
+@import (less) 'codemirror/addon/dialog/dialog.css';
 
 .codeEditor{
   .CodeMirror-foldmarker {
     font-family: inherit;
     text-shadow: none;
     font-weight: 600;
   }
+
+  .cm-tab {
+    background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAADAAAAAMCAQAAACOs/baAAAARUlEQVR4nGJgIAG8JkXxUAcCtDWemcGR1lY4MvgzCEKY7jSBjgxBDAG09UEQzAe0AMwMHrSOAwEGRtpaMIwAAAAA//8DAG4ID9EKs6YqAAAAAElFTkSuQmCC) no-repeat right;
+  }
+
+  .cm-trailingspace {
+    background: url(data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAAgAAAAQAgMAAABW5NbuAAAACVBMVEVHcEwAAAAAAAAWawmTAAAAA3RSTlMAPBJ6PMxpAAAACXBIWXMAAC4jAAAuIwF4pT92AAAAFUlEQVQI12NgwACcCQysASAEZGAAACMuAX06aCQUAAAAAElFTkSuQmCC) repeat-x left;
+    background-size: 7.16px 15px;
+
+    .cm-tab {
+      background: none;
+    }

```

Comment:
This looks good, although I would prefer if tabs still show as arrows.

Do you think we could use something like `specialChars` to add a special rule just for whitespaces? https://discuss.codemirror.net/t/zero-width-spaces-and-styling-of-white-spaces/2054

Perhaps better would be to use the `overlay` system, which is what some other sites uing

Reply:
Yeah I can play around with that and see if it works better

Reply:
FYI I pasted the wrong link to the overlay example. Fixed now.

Reply:
So I've got this working, but as it currently exists it works on all spaces, not just ones that are trailing the line. Do you have a preference between those two?

Reply:
I think we can just apply the CSS to spaces inside the trailing whitespace. That will keep documents looking cleaner.

Reply:
So this does not appear to be viable. We effectively have two options from what I can tell:
1. We keep the CSS here, even though it is admittedly a bit odd.
2. We introduce the specialChars setup to add in the dots on trailing spaces, BUT we accept that there will be random dots littered throughout documents in places we wouldn't expect them. The part that finds/replaces things that match the specialChars regex apparently happens _after_ it has been partially rendered, thus causing certain spaces to have dots in them even though they aren't at the end of the line.

Example: 
With the regex `/ +$/`, all trailing spaces will have `•` in them. However, `This is a **test**` will show up as `This is a•**test**`, because `**test**` being bolded is causing it to be rendered in a different html element, thus meaning that the space after `a` is _technically_ at the end of the line, even though it's not when you look at the editor.

Reply:
What I meant is we handle the display with the CSS rules. So, rather than `/ +$/`, we can just use `/ /` to capture *all* spaces. Then, in the CSS, we use

```
.cm-trailingspace {
    .cm-space {
      background: DOTS;
    }
}
```

So *every* space is a `specialChar`, but only ones inside `.cm-trailingspace` will have the display changed.

Reply:
🤦

Reply:
Yeah that worked lol